codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java

```diff
@@ -70,6 +70,7 @@ public void validateName(CharSequence name) {
     };
 
     private HeaderEntry<CharSequence, CharSequence> firstNonPseudo = head;
+    private final boolean keepInsertionOrder;
```

Comment:
I would prefer to not have this in production code if it is only to support testing. I would prefer to isolate this to the unit tests if possible.

Reply:
Yeah, wasn't sold on that either. 

Other solutions are:
* Duplicate DefaultHttp2Headers in test code
* Remove `final` on `DefaultHttp2Headers#newHeaderEntry` and extends it in test to implement an in insertion order iterator.

Any preference?

Reply:
@madgnome I thin I would go with the "removing final".

Reply:
Removing final sounds OK to me ... the question is how much would we have to duplicate in the tests? It sounds like what you want todo is different than what `DefaultHttp2Headers` is designed todo (enforce HTTP/2 constraints). `DefaultHttp2Headers` also doesn't provide much code besides the `Http2HeaderEntry` (which it looks like would be replaced anyways) and `HTTP2_NAME_VALIDATOR` (which is package private and maybe we can share?). So maybe duplication isn't too bad here?

Reply:
I've changed the implementation of `DefaultHttp2HeadersDecoder` to instantiate `DefaultHttp2Headers` with `keepInsertionOrder` to `true`, doing that allows us to move all the validation to `DefaultHttp2HeadersDecoder` and to keep `HPackDecoder` untouched.

If you think `keepInsertionOrder` still shouldn't exist, I could implement a `DecoderHttp2Headers` class that would be use by the decoder to do that.

Reply:
IIUC `keepInsertionOrder` is still only necessary for tests. If this is the case I would prefer to remove it and isolate this behavior to unit tests.

Reply:
Oh damn, good catch. I was a bit too eager cleaning some code before commiting and remove the important part

```
protected Http2Headers newHeaders() {
        return new DefaultHttp2Headers(validateHeaders, (int) headerArraySizeAccumulator, true);
    }
```

`DefaultHttp2HeadersDecoder` will use `keepInsertionOrder`

Reply:
ping me when behavior has been isolated to unit tests and I will take another look

Reply:
You can see there https://github.com/netty/netty/pull/7500/files#diff-32bd1e577a26442a725a78acaf5822daR163 how we use `keepInsertionOrder` in prod code.

(Tests are back green)

Reply:
@Scottmitch are you happy with `keepInsertionOrder` in `DefaultHttp2Headers` now that it's used by production code?

Otherwise I'll create an `DecoderHttp2Headers` extending `DefaultHttp2Headers` which keep insertion order and use that in the decoder.